benches/hashmap_benchmark.rs

```diff
@@ -0,0 +1,284 @@
+        b.iter_batched(
+            || prepare_struct_insert(),
+            |(ref mut map, ref mut list)| insert_struct_vec(map, list, true),
+            BatchSize::SmallInput,
+        )
+    });
+}
+
+pub fn remove_empty_value_rollback(c: &mut Criterion) {
+    c.bench_function("remove_empty_value_rollback", |b| {
+        b.iter_batched(
+            || prepare_remove(()),
+            |(ref mut map, ref list)| remove_vec(map, list, false),
+            BatchSize::SmallInput,
+        )
+    });
+}
+
+pub fn remove_empty_value_commit(c: &mut Criterion) {
+    c.bench_function("remove_empty_value_commit", |b| {
+        b.iter_batched(
+            || prepare_remove(()),
+            |(ref mut map, ref list)| remove_vec(map, list, true),
+            BatchSize::SmallInput,
+        )
+    });
+}
+
+pub fn remove_struct_value_no_read_rollback(c: &mut Criterion) {
+    c.bench_function("remove_struct_value_no_read_rollback", |b| {
+        b.iter_batched(
+            || prepare_remove(Struct::default()),
+            |(ref mut map, ref list)| remove_vec(map, list, false),
+            BatchSize::SmallInput,
+        )
+    });
+}
+
+pub fn remove_struct_value_no_read_commit(c: &mut Criterion) {
+    c.bench_function("remove_struct_value_no_read_commit", |b| {
+        b.iter_batched(
+            || prepare_remove(Struct::default()),
+            |(ref mut map, ref list)| remove_vec(map, list, true),
+            BatchSize::SmallInput,
+        )
+    });
+}
+
+pub fn search_empty_value(c: &mut Criterion) {
+    c.bench_function("search_empty_value", |b| {
+        b.iter_batched(
+            || prepare_search(()),
+            |(ref map, ref list)| search_vec(map, list),
+            BatchSize::SmallInput,
+        )
+    });
+}
+
+pub fn search_struct_value(c: &mut Criterion) {
+    c.bench_function("search_struct_value", |b| {
+        b.iter_batched(
+            || prepare_search(Struct::default()),
+            |(ref map, ref list)| search_vec(map, list),
+            BatchSize::SmallInput,
+        )
+    });
+}
+
+criterion_group!(insert, insert_empty_value_rollback, insert_empty_value_commit, insert_struct_value_rollback, insert_struct_value_commit);
+criterion_group!(remove, remove_empty_value_rollback, remove_empty_value_commit, remove_struct_value_no_read_rollback, remove_struct_value_no_read_commit);
+criterion_group!(search, search_empty_value, search_struct_value);
+criterion_main!(insert, remove, search);
+
+
+// Utility functions:
+
+fn insert_vec(map: &mut HashMap<u32, ()>, list: &Vec<u32>, commit: bool) {
+    let mut write_txn = map.write();
+    for i in list.iter() {
+        write_txn.insert(*i, ());
+    }
+    if commit { write_txn.commit(); }
+}
+
+fn insert_struct_vec(map: &mut HashMap<u32, Struct>, list: &mut Vec<(u32, StructValue)>, commit: bool) {
+    let mut write_txn = map.write();
+    for i in list.iter_mut() {
+        write_txn.insert(i.0, mem::replace(&mut i.1, None).unwrap());
```

Comment:
Not sure about this mem::replace. You use "ref list" in the closure, but could you make `list: Vec<>`, then use into_iter() instead?

Reply:
I did it to get as little overhead here as I could. Wouldn't this impact performance and increase the error in the test?

Reply:
No. Infact mem::replace will have more overhead here because you have to update the list as well. Rust may "move" the data, but it really just has to copy a certain part of the struct out of the vec then free the vec after use. Where as with mem replace you have to move and rewrite as you go, so into_iter() will actually be faster.

It's weird, but sometimes the "simple" thing in rust is also the "optimal" thing. 

But you know. There is also a way to test this. If only there was some way to benchmark and get evidence about these ideas .... :)